core/unit/name.go

```diff
@@ -0,0 +1,60 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package unit
+
+import (
+	"fmt"
+	"regexp"
+
+	"github.com/juju/juju/internal/errors"
+)
+
+const (
+	// applicationSnippet is a non-compiled regexp that can be composed with
+	// other snippets to form a valid application regexp.
+	applicationSnippet = "(?:[a-z][a-z0-9]*(?:-[a-z0-9]*[a-z][a-z0-9]*)*)"
```

Comment:
**suggestion:** An explanation in english of a regexp is always a must.

=> If it is not bugged, people will prefers to read the explanation
=> if it is, fixer will want to know what was the intention of the regexp.

Moreover this pattern looks suspicious to me:

- a lowercase letter
- zero or more lowercase or digit
- zero or more hyphen followed by at least on lowercase letter surrounded by zero or more lowercase letter or digit.

so `potatoes26-6o6` is ok but `potatoes26-666` isn't ?

Reply:
This is same regex we've been using for 8 years at least, so its defo correct. But I agree about the need for a comment explanation

Reply:
The reason it's a little awkward is we will often a hypen, followed by a number to applications to denote an application's unit. Most of the time we use `/` to separate, but for tags and the like we transform slashes to hypens.

So `potatoes-666` is invalid because there's the potential of ambiguity between the application `potatoes-666` and the 666th unit of the application `potatoes`

Reply:
Strictly, I think we're ok, but we may not have been 8 years ago when this regex was written inititally

Reply:
> This is same regex we've been using for 8 years at least, so its defo correct. But I agree about the need for a comment explanation

Yeah, that's not a valid argument. I am sure there is bug introduced by code that is older than 8 years, in juju and in any program :smile_cat: 

Thanks for the explanation and the comment !